net: sockets: Update setsockopt to handle IPV6_V6ONLY

tedd-an opened the pull request:
This patch removes the IPV6_V6ONLY option since the current socket
implementation doesn't support this option.

Fix Bugs: #14657

Signed-off-by: Tedd Ho-Jeong An <tedd.an@intel.com>

tedd-an commented:
I don't know the story of IPV6_V6ONLY but it looks like it is not supported by the current implementation (maybe it was supported before??)

Even if it wasn't supported and returned an error, it was never a problem because it didn't check the return value until recent changes 19ed83c, then it started to fail.

Another way to fix I can think is simply adding this flag to setsockopt() and do nothing like TCP_NODELAY but I am not sure whether that's a good way to do it.
If we don't support it, it is better to be removed...

Let me know.

pfalcon reviewed (changes requested):
Please don't remove code implemented by fellow developers. Thanks.

Now try to build this sample on a POSIX system, e.g. Linux, and behold the breakage it introduces.

pfalcon commented:
> Another way to fix I can think is simply adding this flag to setsockopt() and do nothing like TCP_NODELAY

Yes, that's the right way to fix it. IPV6_V6ONLY is the default behavior of Zephyr IP stack.

Also mind that socket samples are in general intended to be portable across POSIX systems (of which Zephyr is an aspiring neophyte), so any sample which has `Makefile.posix`, on any change should be built and run on a POSIX system, e.g. Linux. That's actually the way to validate socket samples - how to we know if a particular sample behaves properly? It does, if it behaves the same way as on Linux.

tedd-an committed:
net: sockets: Update setsockopt to handle IPV6_V6ONLY

This patch adds a routine that handles IPV6_V6ONLY option in setsockopt
function.

Fixes #14657

Signed-off-by: Tedd Ho-Jeong An <tedd.an@intel.com>

pfalcon reviewed (approved):
Thanks!

jukkar commented:
> That's actually the way to validate socket samples - how to we know if a particular sample behaves properly? It does, if it behaves the same way as on Linux.

That can be argued. I would say it is one way to validate that the socket sample works. Linux compatibility is nice to have but by no means mandatory. After all we are making Zephyr samples here, and not Linux ones.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1133,6 +1133,16 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
 			return 0;
```
Please use "Fixes #xxxxx" keyword in the commit message. According to https://help.github.com/en/articles/closing-issues-using-keywords the "Fix" might work here but in all the other places we have "Fixes" string.

tedd-an replied:
This is good to know. Thanks!

pfalcon commented:
> I would say it is one way to validate that the socket sample works.

Sure, just one way.

> After all we are making Zephyr samples here, and not Linux ones.

And in Zephyr, we implement the POSIX standard, which is also implemented by other systems, like Linux. So, a good measure (one of!) how well we implement it is to see if the code actually runs portability. (And besides, we implement POSIX not just as an aim in itself, but exactly to allow to reuse code already written [for other systems]).
